### PR TITLE
Freyja subworkflow with required modules

### LINKED ISSUE #3002: new modules & subworkflow: Freyja - variants, update,  demix, bootstrap
### Is there an existing subworkflow for this?

- [X] I have searched for the existing subworkflow

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this subworkflow, I added myself to the `Assignees` to facilitate tracking who is working on the subworkflow

### PR DESCRIPTION
**Pull request by Joon-Klaps**
The [Freyja tool](https://github.com/andersen-lab/Freyja), is a tool to recover relative lineage abundances from mixed SARS-CoV-2 samples from a sequencing dataset, i.e. wastewater. I'm considering implementing this in a pipeline and am sharing my used code.

The PR consists of 
-  a new subworkflow  bam_variant_demix_boot_freyja
- required modules:
  - freyja/variants
  - freyja/demix
  - freyja/update
  - freyja/boot

No md5sums for module tests were kept as some results were not constant across containers or depend on the UShER database.

## PR checklist

Closes #3002 
- [X] This comment contains a description of changes (with reason).
- [X] If you've fixed a bug or added code that should be tested, add tests! (subworkflow tests are all being skipped)
- [X] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [X] Emit the `versions.yml` file.
- [X] Follow the naming conventions.
- [X] Follow the parameters requirements.
- [X] Follow the input/output options guidelines.
- [X] Add a resource `label`
- [X] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [X] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [X] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by nvnieuwk** [comment]
Hi, is it possible to split this PR to one per module/subworkflow? This makes it easier to review and will make sure your PR actually gets reviewed (no one really wants to review a PR with 26 new files 😁)

**Comment by Joon-Klaps** [comment]
I guess it's possible but I'm not certain it would make things easier as the modules `freyja/demix` and `freyja/boot`depend on `freyja/variants`and `freyja/update` and it's also how I wrote the tests for these modules. Submitting them individually would mean waiting for approval of `freyja/variants`and `freyja/update` and subsequently make a PR for all the others which seems cumbersome. Given that the tests are basically a mini subworkflow I thought it would be easier to then also include an actual subworkflow with documentation for it.

**Comment by nvnieuwk** [comment]
Yes I agree, but it's easier for the reviewers to review a smaller PR than a big one 😁. I'll have a look at it soon, but please separate them next time? :)

File: modules/nf-core/freyja/boot/main.nf
Comment (nvnieuwk) [inline modules/nf-core/freyja/boot/main.nf:15]: Can you also add a meta to these inputs?

File: modules/nf-core/freyja/boot/main.nf
Comment (nvnieuwk) [inline modules/nf-core/freyja/boot/main.nf:47]: Can you add a stub section to all modules?

File: modules/nf-core/freyja/boot/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/freyja/boot/meta.yml:66]:
```suggestion
  - lineages:
      type: file
      description: a csv file that includes the lineages present and their corresponding abundances
      pattern: "*.lineages.csv"
  - summarized:
      type: file
      description: a csv file that includes the lineages present but summarized by constellation and their corresponding abundances
      pattern: "*.summarized.csv"
  - versions:
      type: file
      description: File containing software versions
      pattern: "versions.yml"
```

File: modules/nf-core/freyja/demix/main.nf
Comment (nvnieuwk) [inline modules/nf-core/freyja/demix/main.nf:14]: Same here for the meta :)

File: modules/nf-core/freyja/demix/main.nf
Comment (nvnieuwk) [inline modules/nf-core/freyja/demix/main.nf:30]: It's better to not enforce filenames on the user (only file extensions should be hardcoded). If a file name collision is possible you can add a check (see [here](https://nf-co.re/docs/contributing/modules#naming-conventions) part 6)

File: modules/nf-core/freyja/demix/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/freyja/demix/meta.yml:23]: Can you also add all other metas and specify in their description to which file they belong?

File: modules/nf-core/freyja/demix/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/freyja/demix/meta.yml:50]:
```suggestion
  - demix:
      type: file
      description: a tsv file that includes the lineages present, their corresponding abundances, and summarization by constellation
      pattern: "*.demix.tsv"
```

File: modules/nf-core/freyja/update/main.nf
Comment (nvnieuwk) [inline modules/nf-core/freyja/update/main.nf:22]: There are no inputs or arguments possible for this process?

File: modules/nf-core/freyja/variants/main.nf
Comment (nvnieuwk) [inline modules/nf-core/freyja/variants/main.nf:12]: Can you add a meta field to the fasta here?

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
Some change suggestions to the modules, I'll have a look at the subwf if these are fixed (please ping me when it's done 😄)

**Comment by Joon-Klaps** [comment]
@nf-core-bot fix linting

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (nvnieuwk) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:14]:
```suggestion
    ch_bam          // channel: [ val(meta), path(bam) ]
    ch_fasta        // channel: [ val(meta), path(fasta) ]
    val_repeats     // value repeats
    val_db_name     // string db_name
    barcodes        // channel: [ val(meta), path(barcodes) ]
    lineage_meta    // channel: [ val(meta), path(lineage_meta) ]
```

Please also add the meta value to the modules using barcodes and lineage meta :)

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (nvnieuwk) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:36]:
```suggestion
    if(!barcodes || !lineage_meta){
```

No need to check for the contents. Nextflow will automatically return `false` if an input is `[]`

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (nvnieuwk) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:31]:
```suggestion
    ch_versions=ch_versions.mix(FREYJA_VARIANTS.out.versions.first())
```

Can you add this to all versions mix channels? You only need one versions file per run for each module :)

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (nvnieuwk) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:24]:
```suggestion
```

No need to specify these as those will always be created in the subworkflow

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (nvnieuwk) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:39]:
```suggestion
        ch_barcodes=FREYJA_UPDATE.out.barcodes
        ch_lineage_meta=FREYJA_UPDATE.out.lineages_meta
```

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (nvnieuwk) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:73]:
```suggestion
    variants    = FREYJA_VARIANTS.out.variants  // channel: [ val(meta), path(variants_tsv) ]
    depths      = FREYJA_VARIANTS.out.depths    // channel: [ val(meta), path(depths_tsv) ]
    demix       = FREYJA_DEMIX.out.demix        // channel: [ val(meta), path(demix_tsv) ]
    lineages    = FREYJA_BOOT.out.lineages      // channel: [ val(meta), path(lineages_csv) ]
    summarized  = FREYJA_BOOT.out.summarized    // channel: [ val(meta), path(summarized_csv) ]
    versions    = ch_versions                   // channel: [ path(versions.yml) ]
```

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/meta.yml
Comment (nvnieuwk) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/meta.yml:14]: Can you update your meta file to follow the guidelines? It should contains all input and output channels/values with the structure of those channels. See [here](https://nf-co.re/docs/contributing/subworkflows#documentation) for an example

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
Almost done for the modules, here is the subworkflow review. Looked very good, just some documentation issues, I know most subwfs don't do it right either, but we're in the process of fixing that :)

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (Joon-Klaps) [reply subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:36]: I didn't knew this, that's cool!

File: modules/nf-core/freyja/demix/main.nf
Comment (nvnieuwk) [reply modules/nf-core/freyja/demix/main.nf]: You could loop over all those files and check them individually. I've done it before (I'll see if I can find it)

File: modules/nf-core/freyja/update/main.nf
Comment (nvnieuwk) [reply modules/nf-core/freyja/update/main.nf]: Now I can see them :) Any more questions on this?

File: modules/nf-core/freyja/boot/main.nf
Comment (nvnieuwk) [reply modules/nf-core/freyja/boot/main.nf]: I can't seem to find them :/ Weird that this has happened twice now

**Review by Joon-Klaps — COMMENTED** [review]
Think, I've addressed everything. If I missed something lemme know

**Comment by nvnieuwk** [comment]
Yes, you still need to add a `meta` to `barcodes` and `lineages` input files in some modules and adjust your subwf according to that :)

**Review by nvnieuwk — APPROVED** [review]
LGTM! For subworkflows it's a sort of unwritten rule that you need two approvals. So please post a link on Slack too for a second review :) (There is a `#request-review` channel)

File: modules/nf-core/freyja/boot/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/boot/main.nf:15]: Wondering why we need `meta` for all of these inputs @nvnieuwk ? Are `depths`, `barcodes` and `lineages` tied to individual samples or will they be the same across all samples fed to the pipeline @Joon-Klaps ? If the latter, then I think we can remove the `meta`. Just means unnecessary channel manipulations will be required.

Comment (drpatelh) [reply modules/nf-core/freyja/boot/main.nf:15]: Same comment for other modules added in this PR.

File: modules/nf-core/freyja/boot/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/boot/main.nf:20]:
```suggestion
    tuple val(meta), path("*lineages.csv")  , emit: lineages
    tuple val(meta), path("*summarized.csv"), emit: summarized
    path "versions.yml"                     , emit: versions
```

File: modules/nf-core/freyja/demix/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/demix/main.nf:18]:
```suggestion
    tuple val(meta), path("*.tsv"), emit: demix
    path "versions.yml"           , emit: versions
```

File: modules/nf-core/freyja/boot/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/boot/main.nf:15]:
```suggestion
    tuple val(meta3), path(barcodes)
    tuple val(meta4), path(lineages_meta)
```

File: modules/nf-core/freyja/demix/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/demix/main.nf:14]:
```suggestion
    tuple val(meta3), path(barcodes)
    tuple val(meta4), path(lineages_meta)
```

File: modules/nf-core/freyja/update/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/update/main.nf:10]:
```suggestion

    val db_name
```

File: modules/nf-core/freyja/update/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/update/main.nf:16]:
```suggestion
    path "${db_name}/usher_barcodes.csv"   , emit: barcodes
    path "${db_name}/lineages.yml"         , emit: lineages_topology
    path "${db_name}/curated_lineages.json", emit: lineages_meta
    path "versions.yml"                    , emit: versions
```

File: modules/nf-core/freyja/update/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/update/main.nf:25]:
```suggestion
    mkdir -p $db_name
```

File: modules/nf-core/freyja/variants/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/variants/main.nf:25]:
```suggestion
```

File: modules/nf-core/freyja/boot/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/boot/main.nf:28]:
```suggestion
```

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (drpatelh) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:24]:
```suggestion
    FREYJA_VARIANTS (
        ch_bam,
        ch_fasta
    )
```

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (drpatelh) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:28]:
```suggestion
    ch_versions = ch_versions.mix(FREYJA_VARIANTS.out.versions.first())
```

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (drpatelh) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:44]:
```suggestion
    if (!ch_barcodes || !ch_lineages_meta) {
        FREYJA_UPDATE (
            val_db_name
        )
        ch_versions = ch_versions.mix(FREYJA_UPDATE.out.versions.first())

        FREYJA_UPDATE
            .out
            .barcodes
            .map { barcodes  -> [ [], barcodes ] }
            .set { ch_barcodes }

        FREYJA_UPDATE
            .out
            .lineages_meta
            .map { lineages  -> [ [], lineages ] }
            .set { ch_lineages_meta }
    }
```

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (drpatelh) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:55]:
```suggestion
    FREYJA_DEMIX (
        ch_freyja_variants,
        ch_freyja_depths,
        ch_barcodes,
        ch_lineages_meta
    )
    ch_freyja_demix = FREYJA_DEMIX.out.demix
    ch_versions = ch_versions.mix(FREYJA_DEMIX.out.versions.first())
```

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (drpatelh) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:67]:
```suggestion
    FREYJA_BOOT (
        ch_freyja_variants,
        ch_freyja_depths,
        val_repeats,
        ch_barcodes,
        ch_lineages_meta
    )
    ch_versions = ch_versions.mix(FREYJA_BOOT.out.versions.first())
```

File: subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf
Comment (drpatelh) [inline subworkflows/nf-core/bam_variant_demix_boot_freyja/main.nf:77]:
```suggestion
    variants       = FREYJA_VARIANTS.out.variants  // channel: [ val(meta), path(variants_tsv) ]
    depths         = FREYJA_VARIANTS.out.depths    // channel: [ val(meta), path(depths_tsv) ]
    demix          = FREYJA_DEMIX.out.demix        // channel: [ val(meta), path(demix_tsv) ]
    lineages       = FREYJA_BOOT.out.lineages      // channel: [ val(meta), path(lineages_csv) ]
    summarized     = FREYJA_BOOT.out.summarized    // channel: [ val(meta), path(summarized_csv) ]
    barcodes       = ch_barcodes                   // channel: [ val(meta), path(barcodes) ]
    lineages_meta  = ch_lineages_meta              // channel: [ val(meta), path(lineages_meta) ]
    versions       = ch_versions                   // channel: [ path(versions.yml) ]
```

**Review by drpatelh — CHANGES_REQUESTED** [review]
Great work @Joon-Klaps ! Mostly syntax consistency comments that should be solved on batch commit. Was looking for exactly this to fix https://github.com/nf-core/viralrecon/issues/299 🚀 

As you mentioned on Slack:
"Another feauture that I might think be interesting is the implementation of wastewatercovid spades (upcoming version of spades will have this is my guess) or the freyja pipeline for waste water covid analyses. I’ve created a subworkflow for Freyja https://github.com/nf-core/modules/pull/3003#pullrequestreview-1336437972 and for the time being a version of viralrecon with freyja implemented to see how it would go https://github.com/Joon-Klaps/viralrecon/tree/waste-water-covid."

File: modules/nf-core/freyja/boot/main.nf
Comment (nvnieuwk) [reply modules/nf-core/freyja/boot/main.nf:15]: That was something we agreed on with the maintainers 🤷. I agree it doesn't make sense in some cases but guidelines are guidelines? 😁

File: modules/nf-core/freyja/boot/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/boot/main.nf:15]:
```suggestion
    tuple val(meta),  path(variants), path(depths)
    path barcodes
    path lineages_meta
    val repeats
```

File: modules/nf-core/freyja/demix/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/demix/main.nf:14]:
```suggestion
    tuple val(meta), path(variants), path(depths)
    path barcodes
    path lineages_meta
```

File: modules/nf-core/freyja/update/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/update/main.nf:9]:
```suggestion

    input:
```

File: modules/nf-core/freyja/variants/main.nf
Comment (drpatelh) [inline modules/nf-core/freyja/variants/main.nf:12]:
```suggestion
    tuple val(meta), path(bam)
    path fasta
```

**Review by drpatelh — APPROVED** [review]
Awesome! 🚀

**Pull request merged**
